airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,482 @@
+        :param weight: number of requests to deduct from credit
+        :return:
+        """
+
+    @abc.abstractmethod
+    def update(self, available_calls: Optional[int], call_reset_ts: Optional[datetime.datetime]) -> None:
+        """Update call rate counting with current values
+
+        :param available_calls:
+        :param call_reset_ts:
+        """
+
+
+class RequestMatcher(abc.ABC):
+    """Callable that help to match a request object with call rate policies."""
+
+    @abc.abstractmethod
+    def __call__(self, request: Any) -> bool:
+        """
+
+        :param request:
+        :return: True if pattern matches the provided request object, False - otherwise
+        """
+
+
+class HttpRequestMatcher(RequestMatcher):
+    """Simple implementation of RequestMatcher for http requests case"""
+
+    def __init__(
+        self,
+        method: Optional[str] = None,
+        url: Optional[str] = None,
+        params: Optional[Mapping[str, Any]] = None,
+        headers: Optional[Mapping[str, Any]] = None,
+    ):
+        """Constructor
+
+        :param method:
+        :param url:
+        :param params:
+        :param headers:
+        """
+        self._method = method
+        self._url = url
+        self._params = {str(k): str(v) for k, v in (params or {}).items()}
+        self._headers = {str(k): str(v) for k, v in (headers or {}).items()}
+
+    @staticmethod
+    def _match_dict(obj: Mapping[str, Any], pattern: Mapping[str, Any]) -> bool:
+        """Check that all elements from pattern dict present and have the same values in obj dict
+
+        :param obj:
+        :param pattern:
+        :return:
+        """
+        return pattern.items() <= obj.items()
+
+    def __call__(self, request: Any) -> bool:
+        """
+
+        :param request:
+        :return: True if pattern matches the provided request object, False - otherwise
+        """
+        if isinstance(request, requests.Request):
+            prepared_request = request.prepare()
+        elif isinstance(request, requests.PreparedRequest):
+            prepared_request = request
+        else:
+            return False
+
+        if self._method is not None:
+            if prepared_request.method != self._method:
+                return False
+        if self._url is not None and prepared_request.url is not None:
+            url_without_params = prepared_request.url.split("?")[0]
+            if url_without_params != self._url:
+                return False
+        if self._params is not None:
+            parsed_url = parse.urlsplit(prepared_request.url)
+            params = dict(parse.parse_qsl(str(parsed_url.query)))
+            if not self._match_dict(params, self._params):
+                return False
+        if self._headers is not None:
+            if not self._match_dict(prepared_request.headers, self._headers):
+                return False
+        return True
+
+
+class BaseCallRatePolicy(AbstractCallRatePolicy, abc.ABC):
```

Comment:
@brianjlai FYI since you were asking about supporting different types of rate limits. 

We should only need FixedWindow for Stripe, but the [moving window strategy](https://github.com/airbytehq/airbyte/pull/31276/files#diff-e21ba47ba176b1cf99b7ad8c7b003847b710bde1082561515560928c4275a8f0R245) shows how we can implement more complex logic